### PR TITLE
[google_maps_flutter_android] Convert `PlatformCameraUpdate` to pigeon.

### PR DESCRIPTION
**Pull request by yaakovschectman**
Because this also converts the platform interface Camera type, this will need to be a federated change, unless we come up with a better idea of how to convert that class.

https://github.com/flutter/flutter/issues/152928

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java [hunk 1]
```diff
@@ -339,37 +339,50 @@ static CameraPosition toCameraPosition(Object o) {
...
-        throw new IllegalArgumentException("Cannot interpret " + o + " as CameraUpdate");
+  static CameraUpdate cameraUpdateFromPigeon(Messages.PlatformCameraUpdate update, float density) {
+    Messages.PlatformNewCameraPosition newCameraPosition = update.getNewCameraPosition();
+    if (newCameraPosition != null) {
+      return CameraUpdateFactory.newCameraPosition(
+          cameraPositionFromPigeon(newCameraPosition.getCameraPosition()));
+    }
+    Messages.PlatformNewLatLng newLatLng = update.getNewLatLng();
+    if (newLatLng != null) {
+      return CameraUpdateFactory.newLatLng(latLngFromPigeon(newLatLng.getLatLng()));
```
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:350]: I may not be following the types here but is it possible to make this CameraUpdateFactory.newLatLng(latLngFromPigeon(newLatLng); It seems, based on names alone that we should not need so many unboxing/reboxing of lat long data.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java [hunk 2]
```diff
@@ -339,37 +339,50 @@ static CameraPosition toCameraPosition(Object o) {
...
-        return CameraUpdateFactory.zoomIn();
-      case "zoomOut":
-        return CameraUpdateFactory.zoomOut();
-      case "zoomTo":
-        return CameraUpdateFactory.zoomTo(toFloat(data.get(1)));
-      default:
-        throw new IllegalArgumentException("Cannot interpret " + o + " as CameraUpdate");
+  static CameraUpdate cameraUpdateFromPigeon(Messages.PlatformCameraUpdate update, float density) {
+    Messages.PlatformNewCameraPosition newCameraPosition = update.getNewCameraPosition();
+    if (newCameraPosition != null) {
```
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:344]: I see how this code is a conversion from what was here before but I think the if statement make it harder to read what should happen. Before the toCameraUpdate method either updated the camera position, the lat/lng or the bounds. Now that is much harder to understand. 

Try restructuring this method to make it easier for a human reader and/or adding java doc comments.

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart [hunk 3]
```diff
@@ -796,6 +796,48 @@ class GoogleMapsFlutterAndroid extends GoogleMapsFlutterPlatform {
     return PlatformTileOverlay(
         json: tileOverlay.toJson() as Map<String, Object?>);
   }
+
+  static PlatformCameraUpdate _platformCameraUpdateFromCameraUpdate(
+      CameraUpdate update) {
+    if (update is NewCameraPosition) {
```
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart:802]: You might have discussed this with @stuartmorgan if so feel free to resolve but can cameraUpdate include an enum of the type of update instead of these long object type checks?

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart [hunk 4]
```diff
@@ -155,49 +147,131 @@ class CameraUpdate {
...
   static CameraUpdate zoomTo(double zoom) {
-    return CameraUpdate._(<Object>['zoomTo', zoom]);
+    return ZoomTo(zoom);
   }
 
-  final Object _json;
-
   /// Converts this object to something serializable in JSON.
-  Object toJson() => _json;
+  Object toJson();// => _json;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart:182]: The commented out code should be removed.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart [hunk 5]
```diff
@@ -155,49 +147,131 @@ class CameraUpdate {
...
+}
+
+/// Defines a camera move to a new position.
+class NewCameraPosition extends CameraUpdate {
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart:186]: I think it would probably be easier to keep track of all of these classes if they had a `CameraUpdate` prefix.

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart (hunk 3 above)
Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart:802]: This was discussed a bit for the Pigeon classes (where I don't think the enum probably helps us), but not for the Dart code.

I think for Dart, where we will be creating a new permanent public API surface, it would be better to add a type enum, and to make all of these classes private, so that we're not locking ourselves into a specific class structure forever.

File: packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart [hunk 6]
```diff
@@ -40,17 +40,79 @@ class PlatformCameraPosition {
...
+
+  final PlatformNewCameraPosition? newCameraPosition;
+  final PlatformNewLatLng? newLatLng;
+  final PlatformNewLatLngBounds? newLatLngBounds;
+  final PlatformNewLatLngZoom? newLatLngZoom;
+  final PlatformScrollBy? scrollBy;
+  final PlatformZoomBy? zoomBy;
+  final bool zoomIn;
+  final bool zoomOut;
+  final PlatformZoomTo? zoomTo;
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: This isn't the design we discussed in the issue; what's the reason for the change?

Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: If the change you are referring to is `PlatformCameraUpdate` containing members instead of each option extending it as a base class, this is due to the limitations of Pigeon, as we discussed privately. If each of the individual subclasses extend `PlatformCameraUpdate`, but are not reachable by the exposed APIs, they will not be converted by Pigeon at all, and therefore be unusable. The only other alternative I can think of that still keeps the subtypes reachable is to overload `animate/moveCamera` for each of the subtypes, which does not sound very preferable.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java (hunk 1 above)
Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:350]: `newLatLng` is an instance of `PlatformNewLatLng`, which represents the camera move update to the new position. `latLngFromPigeon` accepts a `PlatformLatLng`, which is just a boxed pair of numeric coordinates, not a `PlatformNewLatLng`. I intend to add a `CameraUpdate` prefix to the relevant classes, including `PlatformNewLatLng`, as per Stuart's comment, which I believe will make this more clear.

File: packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart (hunk 6 above)
Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: Oh, I didn't realize that was the context for that question.

@tarrinneal How hard would it be to generate all classes in the current implementation? I never remember if it works this way because it used to have to when Pigeon used mirrors, and it just stayed that way, or if it had to change to this way when we moved *off* of mirrors.

Comment (tarrinneal) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: I'm not sure I'm understanding the question here. @stuartmorgan

Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: @tarrinneal I believe he is basically asking the following. Suppose there is some pigeon defined class `X` which is exposed via APIs and is therefore converted by the code generation when running pigeon. Is there a way to define multiple subclasses of `X` that will also be converted by code gen without defining new functions that explicitly use them?

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart (hunk 3 above)
Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart:802]: If these classes were private, would that not then make their members inaccessible in this file?

File: packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart (hunk 6 above)
Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: We chatted about this briefly in triage yesterday, and @tarrinneal believes this would be easy to fix in Pigeon, in which case we should probably do that and then update this PR accordingly.

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart (hunk 3 above)
Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart:802]: 🤦🏻 

Sorry, I was thinking about mirroring the Java SDK's approach, but there they only need to consume the properties internally to the SDK.

File: packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart (hunk 6 above)
Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: Okay, so how would we do that? By "fix in Pigeon", do you mean pushing a change to Pigeon itself?

Comment (tarrinneal) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: I'll push the pigeon fix asap

Comment (tarrinneal) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: https://github.com/flutter/packages/pull/7529

Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: @stuartmorgan It sounds like what Tarrin has fixed in Pigeon does not actually resolve this, as inheritance is not supported in Pigeon in any capacity.

Comment (tarrinneal) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: Adding inheritance is a much bigger project I don't think that is something that can ready in a reasonable time frame for this pr.

Comment (reidbaker) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: Could we or do we want to bypass inheritance and say that the cameraUpdate object contains an Object and an a defined enum for the type. Then in methods that take cameraUpdate switch on an enum and cast.

Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:63]: > I think I am still blocked. It looks like Pigeon's generated code does not preserve inheritance between classes

This is why the issue describes using `Object` as the item type. Inheritance would be nice, but shouldn't be blocking.

> Could we or do we want to bypass inheritance and say that the cameraUpdate object contains an Object and an a defined enum for the type. Then in methods that take cameraUpdate switch on an enum and cast.

We don't even need the enum; we can still do type detection on the class without inheritance.

We could do it if you feel strongly, but I think it's just adding duplicate information. Checking the actual type before casting to that type seems much better to me than checking an enum and then casting.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java [hunk 7]
```diff
@@ -340,37 +340,56 @@ static CameraPosition toCameraPosition(Object o) {
...
+    if (cameraUpdate instanceof Messages.PlatformCameraUpdateZoomTo) {
+      Messages.PlatformCameraUpdateZoomTo zoomTo =
+          (Messages.PlatformCameraUpdateZoomTo) cameraUpdate;
+      return CameraUpdateFactory.zoomTo(zoomTo.getZoom().floatValue());
+    }
+    if (cameraUpdate instanceof Messages.PlatformCameraUpdateZoom) {
+      Messages.PlatformCameraUpdateZoom zoom = (Messages.PlatformCameraUpdateZoom) cameraUpdate;
+      return (zoom.getOut()) ? CameraUpdateFactory.zoomOut() : CameraUpdateFactory.zoomIn();
+    }
+    throw new IllegalArgumentException("PlatformCameraUpdate must have one non-null member.");
```
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:392]: This error message does not make sense to me since the throw is not related to a value being null.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java [hunk 8]
```diff
@@ -494,6 +513,14 @@ static Point pointFromPigeon(Messages.PlatformPoint point) {
     return new Point(point.getX().intValue(), point.getY().intValue());
   }
 
+  @Nullable
+  static Point pointFromPigeon(@Nullable Messages.PlatformOffset point, float density) {
+    if (point == null) {
+      return null;
+    }
+    return new Point((int) (point.getDx() * density), (int) (point.getDy() * density));
```
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:521]: why should this be a truncation (cast to int) instead of a round?

**Review by reidbaker — APPROVED** [review]

**Comment by reidbaker** [comment]
Please hold on @stuartmorgan's approval as well.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java (hunk 8 above)
Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:521]: The previous implementation replaced by this PR called `toPoint`, which called `toPixels`, which performed a cast to int. I did not intend this PR to change any implementation details beyond how the data is structured.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java (hunk 7 above)
Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:392]: It made sense before, but now I need to update it.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart [hunk 9]
```diff
@@ -110,42 +110,59 @@ class CameraPosition {
       'CameraPosition(bearing: $bearing, target: $target, tilt: $tilt, zoom: $zoom)';
 }
 
+/// Indicates which type of camera update this instance represents.
+enum CameraUpdateType {
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart:114]: Ideally we would make CameraUpdate a sealed class so that we could just do exhaustive checked switches over the classes themselves, but since that would technically be a breaking change this works.

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart [hunk 10]
```diff
@@ -799,6 +799,60 @@ class GoogleMapsFlutterAndroid extends GoogleMapsFlutterPlatform {
...
+                    : _platformOffsetFromOffset(update.focus!)));
+      case CameraUpdateType.zoomIn:
+        update as CameraUpdateZoomIn;
+        return PlatformCameraUpdate(
+            cameraUpdate: PlatformCameraUpdateZoom(out: false));
+      case CameraUpdateType.zoomOut:
+        update as CameraUpdateZoomOut;
+        return PlatformCameraUpdate(
+            cameraUpdate: PlatformCameraUpdateZoom(out: true));
+      case CameraUpdateType.scrollBy:
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart:849]: This is the only case currently covered by Dart unit tests. It was okay to only test one update type when the details of how they were serialized was up to a different package, but now that there is logic for each case within this package, the unit tests need to be updated so there's coverage of all of these branches.

File: packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart [hunk 11]
```diff
@@ -110,42 +110,59 @@ class CameraPosition {
...
+  zoomIn,
+  /// Zoom out
+  zoomOut,
+}
+
 /// Defines a camera move, supporting absolute moves as well as moves relative
 /// the current position.
-class CameraUpdate {
-  const CameraUpdate._(this._json);
+abstract class CameraUpdate {
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/camera.dart:137]: All of the changes here need Dart unit tests. I would recommend one new test for each factory that validates:
- that the class is as expected,
- that the `updateType` is as expected, and
- that the properties were passed through correctly.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java [hunk 12]
```diff
@@ -340,37 +340,56 @@ static CameraPosition toCameraPosition(Object o) {
     return builder.build();
   }
 
-  static CameraUpdate toCameraUpdate(Object o, float density) {
-    final List<?> data = toList(o);
-    switch (toString(data.get(0))) {
-      case "newCameraPosition":
-        return CameraUpdateFactory.newCameraPosition(toCameraPosition(data.get(1)));
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:347]: toCameraPosition appears to be unused now, so is dead code that should be removed.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java [hunk 13]
```diff
@@ -340,37 +340,56 @@ static CameraPosition toCameraPosition(Object o) {
...
-      case "newLatLng":
-        return CameraUpdateFactory.newLatLng(toLatLng(data.get(1)));
-      case "newLatLngBounds":
-        return CameraUpdateFactory.newLatLngBounds(
-            toLatLngBounds(data.get(1)), toPixels(data.get(2), density));
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:352]: toLatLngBounds is also dead code.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java [hunk 14]
```diff
@@ -340,37 +340,56 @@ static CameraPosition toCameraPosition(Object o) {
...
-        return CameraUpdateFactory.newLatLngZoom(toLatLng(data.get(1)), toFloat(data.get(2)));
-      case "scrollBy":
-        return CameraUpdateFactory.scrollBy( //
-            toFractionalPixels(data.get(1), density), //
-            toFractionalPixels(data.get(2), density));
-      case "zoomBy":
-        if (data.size() == 2) {
-          return CameraUpdateFactory.zoomBy(toFloat(data.get(1)));
-        } else {
-          return CameraUpdateFactory.zoomBy(toFloat(data.get(1)), toPoint(data.get(2), density));
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:363]: toPoint is also dead code, which cascades to toPixels and toFractionalPixels being dead code as well.

I believe if you edit this file in Android Studio, the "problems" pane will flag anything that is `static` and unused.

Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:363]: (As a general note, I would expect that when the Pigeon conversion of `google_maps_flutter_android` is complete, there would be no methods in this file that take an `Object` as their parameter, so it would be good to audit to make sure that dead code is removed, and anything left already has a tracking bug. Per [this comment](https://github.com/flutter/flutter/issues/152926) I'm not sure we have all the remaining work tracked in issues.)

File: packages/google_maps_flutter/google_maps_flutter/example/pubspec.yaml [hunk 15]
```diff
@@ -33,3 +33,8 @@ flutter:
   uses-material-design: true
   assets:
     - assets/
+
+# FOR TESTING AND INITIAL REVIEW ONLY. DO NOT MERGE.
+# See https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changing-federated-plugins
+dependency_overrides:
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/example/pubspec.yaml:39]: The changes to `google_maps_flutter` can be reverted.

**Review by stuartmorgan — COMMENTED** [review]
Some comments, but nothing structural at this point. Feel free to split `google_maps_flutter_platform_interface` out into its own PR and just addressing the testing in that PR.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
(I forgot to switch from Comment to Request Changes to reflect the comments.)

**Comment by yaakovschectman** [comment]
> Some comments, but nothing structural at this point. Feel free to split `google_maps_flutter_platform_interface` out into its own PR and just addressing the testing in that PR.

@stuartmorgan I do not think we can address the testing in the PR that updates the platform interface. The tests would be in the android package, and no PR can land that refers to changes in the platform interface until the platform interface's package's PR lands. Perhaps I have misunderstood what you mean.

File: packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart [hunk 16]
```diff
@@ -40,17 +40,66 @@ class PlatformCameraPosition {
...
-  /// The update data, as JSON. This should only be set from
-  /// CameraUpdate.toJson, and the native code must interpret it according to the
-  /// internal implementation details of the CameraUpdate class.
-  // TODO(stuartmorgan): Update the google_maps_platform_interface CameraUpdate
-  //  class to provide a structured representation of an update. Currently it
-  //  uses JSON as its only state, so there is no way to preserve structure.
-  //  This wrapper class exists as a placeholder for now to at least provide
-  //  type safety in the top-level call's arguments.
-  final Object json;
+  PlatformCameraUpdate({required this.cameraUpdate});
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:43]: Making this a named argument vs a positional argument doesn't seem to add any clarity, and it makes the already-verbose code using this class even more verbose. Having the single argument just be positional would be clearer in this case IMO.

File: packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart [hunk 17]
```diff
@@ -40,17 +40,66 @@ class PlatformCameraPosition {
...
+
+  /// This Object shall be any of the below classes prefixed with
+  /// PlatformCameraUpdate. Each such class represents a different type of
+  /// camera update, and each holds a different set of data, preventing the
+  /// use of a single unified class. Pigeon does not support inheritance, which
+  /// prevents a more strict type bound.
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:49]: You can reference https://github.com/flutter/flutter/issues/117819 here.

File: packages/google_maps_flutter/google_maps_flutter_android/test/google_maps_flutter_android_test.dart [hunk 18]
```diff
@@ -104,7 +104,7 @@ void main() {
     expect(bounds, expectedBounds);
   });
 
-  test('moveCamera calls through', () async {
+  test('moveCamera calls through scrollBy', () async {
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/test/google_maps_flutter_android_test.dart:107]: I'm confused by the naming changes here; the API being called is `moveCamera` at both layers; nothing called `scrollBy` is being called in either method.

You could say "calls through with expected parameters" if you want to be more explicit about what's being verified.

File: packages/google_maps_flutter/google_maps_flutter_android/test/google_maps_flutter_android_test.dart [hunk 19]
```diff
@@ -774,6 +782,164 @@ void main() {
     },
   );
 
+  test('moveCamera calls through newCameraPosition', () async {
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_android/test/google_maps_flutter_android_test.dart:785]: `moveCamera calls though with expected newCameraPosition`; similarly with the methods below.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Thanks for the expanded testing and Java cleanup! Mostly looks great, just a few comments.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java (hunk 8 above)
Comment (reidbaker) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:507]: I believe that this is a logic change. toPixels calls toFractionalPixels which multiplies a float by the density. 
```
  private static float toFractionalPixels(Object o, float density) {
    return toFloat(o) * density;
  }
```

In none of the code I traced did I see us dropping data by an int cast.

**Review by reidbaker — APPROVED** [review]

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java [hunk 20]
```diff
@@ -494,16 +499,16 @@ static Point pointFromPigeon(Messages.PlatformPoint point) {
...
   }
 
-  static Messages.PlatformPoint pointToPigeon(Point point) {
-    return new Messages.PlatformPoint.Builder().setX((long) point.x).setY((long) point.y).build();
-  }
-
-  private static LatLngBounds toLatLngBounds(Object o) {
-    if (o == null) {
+  @Nullable
+  static Point pointFromPigeon(@Nullable Messages.PlatformOffset point, float density) {
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:503]: I believe I established that naming convention in this file. The problem is that during this transition there have been a bunch of cases where there are two different version of all the methods: one that converts method channel primitives, for the unconverted code, and one that converts Pigeon objects.

In that context, the fact that the source object is Pigeon is very important. Especially because the old methods take `Object` which means you can accidentally pass a Pigeon object to them (which will then crash).

I would have no objection to mass renaming these when the conversion is done and the old versions don't exist any more. Or we could mass rename the remaining old methods now to something that clearly identifies them, and then mass rename the Pigeon versions to use the generic names.

[the remark this answers is not in the document]
Comment (reidbaker) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:503]: That makes sense and makes me glad I asked the question.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java (hunk 8 above)
Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:507]: `toPixels` converts the result of `toFractionalPixels` to int via (int) cast: https://github.com/flutter/packages/pull/7507/files#diff-26752280ec885e4465baf81a4771b5a7e6fd1041d333c9fedd712bcc5c3916c7L534.  
`toPoint` calls `toPixels`: https://github.com/flutter/packages/pull/7507/files#diff-26752280ec885e4465baf81a4771b5a7e6fd1041d333c9fedd712bcc5c3916c7L568.  
The `zoomBy` case previously called `toPoint`: https://github.com/flutter/packages/pull/7507/files#diff-26752280ec885e4465baf81a4771b5a7e6fd1041d333c9fedd712bcc5c3916c7L363.  
The new `zoomBy` case calls `pointFromPigeon`: https://github.com/flutter/packages/pull/7507/files#diff-26752280ec885e4465baf81a4771b5a7e6fd1041d333c9fedd712bcc5c3916c7L363.

Comment (reidbaker) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:507]: I am mistaken, resolving.

File: packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart (hunk 16 above)
Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:43]: As far as I can tell, Pigeon will convert positional arguments to named arguments in the output file, anyway. I tried making it positional in `messages.dart`, but after running Pigeon, the generated code does not change. Do you think it would still be useful for this file to use a positional argument?

Comment (stuartmorgan) [reply packages/google_maps_flutter/google_maps_flutter_android/pigeons/messages.dart:43]: Oh, for some I thought we preserved the argument structure on the Dart side. In that case it doesn't matter.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (yaakovschectman) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/Convert.java:503]: @stuartmorgan Do you have any remaining concerns?

**Review by stuartmorgan — APPROVED** [review]
LGTM
